### ISSUE TITLE
Occasional errors during SPI transfers

### ISSUE THREAD

**Issue by samuelfmlourenco** [bloguetronica/cp2130-com#27]
There is still a chance of getting an error, which may be severe enough because the CP2130 remains locked due to a race condition. This is known to be an issue with the CP2130 itself, and provisions to avoid it were added into the code inside "devicewindow.cpp", more specifically inside on_pushButtonRead_clicked(), on_pushButtonWrite_clicked() and on_pushButtonWriteRead_clicked(). Each of these functions has a 100 us pause forced before disabling the chip select, once the transfer is done:

```
usleep(100);  // Wait 100 us, in order to prevent possible errors while disabling the chip select (workaround)
cp2130_.disableCS(channel, errcnt, errstr);  // Disable the previously enabled chip select
```

However, the caveat here is that, due to the implementation of the progress dialogs, these SPI transfer functions are non-blocking, and update() is often called in between partial (bulk) transfers. It happens that update() gets the state of all GPIOs. The CP2130 is also known to fail when trying to manipulate the value of any GPIO right after a bulk transfer. However, it is a possibility that reading GPIOs may also be problematic. This has to be tested, though.

Anyway, the QTimer which is used to call update() could be stopped during a SPI transfer, but that would remove the possibility of seeing updated values during a transfer. There are two more potential ways of solving the issue. One is to rewrite on_pushButtonRead_clicked(), on_pushButtonWrite_clicked() and on_pushButtonWriteRead_clicked(), by simply moving the usleep() call to somewhere inside the while loop of each function, right after the bulk transfer. The other way, less orthodox but ultimately more time efficient, is to add a usleep() inside update(), as follows:

```
// This is the main update routine (expanded in version 3.0)
void DeviceWindow::update()
{
    usleep(100);  // Added line here!
    int errcnt = 0;
    QString errstr;
    quint16 gpios = cp2130_.getGPIOs(errcnt, errstr);
    CP2130::EventCounter evtcntr;
    if (pinConfig_.gpio5 == CP2130::PCEVTCNTRRE || pinConfig_.gpio5 == CP2130::PCEVTCNTRFE || pinConfig_.gpio5 == CP2130::PCEVTCNTRNP || pinConfig_.gpio5 == CP2130::PCEVTCNTRPP) {
        evtcntr = cp2130_.getEventCounter(errcnt, errstr);
    }
    if (opCheck(tr("update-op"), errcnt, errstr)) {  // If no errors occur (the string "update-op" should be translated to "Update")
        updateView(gpios, evtcntr);  // Update values
    }
}
```

Albeit the above mentioned functions are still not "self-correcting", update() is the only function which might be called "concurrently". Also, this accounts for the future possibility of allowing the user to turn off updates during transfers at will, with the purpose of making them more efficient without the added pauses between each partial transfer.

**Comment by samuelfmlourenco** [comment]
Further testing revealed that this issue usually happens during SPI write transfers, the frequency being set at 93.8 kHz. It is rare, but quite repeatable. The code proposed above didn't fix the issue.

**Comment by samuelfmlourenco** [comment]
Rewriting on_pushButtonRead_clicked(), on_pushButtonWrite_clicked() and on_pushButtonWriteRead_clicked(), as proposed above, still didn't fix the issue. The only viable option is to stop the update timer during the SPI transfers.

**Comment by samuelfmlourenco** [comment]
Stopping the update timer during SPI transfers didn't solve the issue either. To investigate this further, it is essential to analyze the USB traffic with Wireshark, and capture an error.

**Comment by samuelfmlourenco** [comment]
This is not very revealing, but it seems that the device just stops responding to bulk transfer requests. The response is ENOENT, which basically means no response from the device.

![cp2130-conf-v30-error](https://user-images.githubusercontent.com/16304376/186762182-00bc395b-86f4-4456-aa8c-7914df411547.png)

**Comment by samuelfmlourenco** [comment]
Could not reproduce the error with the update timer stopped during transfers. Probably, the previous error was a one-off, but this needs to be confirmed. Here is a capture:

![cp2130-conf-v30-noerror](https://user-images.githubusercontent.com/16304376/186765732-68984aee-aff7-45c3-ac66-2abe57ac0c32.png)

**Comment by samuelfmlourenco** [comment]
Managed to reproduce the error previously obtained with the update timer stopped during an SPI write transfer, but only after many tries. As the image below shows, there is no common factor.

![cp2130-conf-v30-error2](https://user-images.githubusercontent.com/16304376/186769248-2ca76048-548c-437c-b8a5-8c1f3bc8df9b.png)

**Comment by samuelfmlourenco** [comment]
Here is another capture of a similar failure. The previous fix didn't solve the issue, but clearly mitigated it. The cause seems to be much more deterministic, as it is evident the existence of a common denominator.

![cp2130-conf-v30-error3](https://user-images.githubusercontent.com/16304376/186772454-f6c4fb97-41aa-4693-ad44-1aef7151ef22.png)

The error could be caused by a timing issue, never seen before. Perhaps, another 100 us pause is needed between enabling the chip select and the first bulk transfer. Like so:

```
// This function was expanded in version 3.0, in order to support transfers greater than 4096 bytes
void DeviceWindow::on_pushButtonRead_clicked()
{
    quint8 channel = static_cast<quint8>(ui->comboBoxChannel->currentText().toUInt());
    size_t fragmentSizeLimit = evaluateSizeLimit();
    size_t bytesToRead = static_cast<size_t>(ui->spinBoxBytesToRead->value());
    size_t bytesProcessed = 0;
    QProgressDialog spiReadProgress(tr("Performing SPI read..."), tr("Abort"), 0, static_cast<int>(bytesToRead), this);  // Progress dialog implemented in version 3.0
    spiReadProgress.setWindowTitle(tr("SPI Read"));
    spiReadProgress.setWindowModality(Qt::WindowModal);
    spiReadProgress.setMinimumDuration(500);  // The progress dialog should appear only if the operation takes more than 500 ms
    Data read;
    timer_->stop();  // Added!
    QElapsedTimer time;
    time.start();
    int errcnt = 0;
    QString errstr;
    cp2130_.selectCS(channel, errcnt, errstr);  // Enable the chip select corresponding to the selected channel, and disable any others
    usleep(100);  // Added!
    while (bytesProcessed < bytesToRead) {
        if (spiReadProgress.wasCanceled()) {  // If the user clicks "Abort"
            break;  // Abort the SPI read operation
        }
        size_t bytesRemaining = bytesToRead - bytesProcessed;
        size_t fragmentSize = bytesRemaining > fragmentSizeLimit ? fragmentSizeLimit : bytesRemaining;
        QVector<quint8> readFragment = cp2130_.spiRead(static_cast<quint32>(fragmentSize), epin_, epout_, errcnt, errstr);  // Read from the SPI bus
        if (errcnt > 0) {  // In case of error
            spiReadProgress.cancel();  // Important!
            break;  // Abort the SPI read operation
        }
        read.vector += readFragment;  // The returned fragment could be considered valid at this point
        bytesProcessed += fragmentSize;
        spiReadProgress.setValue(static_cast<int>(bytesProcessed));
    }
    usleep(100);  // Wait 100 us, in order to prevent possible errors while disabling the chip select (workaround)
    cp2130_.disableCS(channel, errcnt, errstr);  // Disable the previously enabled chip select
    int elapsedTime = static_cast<int>(time.elapsed());  // Elapsed time in milliseconds
    timer_->start();  // Added!
    ui->lineEditRead->setText(read.toHexadecimal());  // At least, a partial result should be shown in case of error
    if (errcnt > 0) {  // Update status bar
        labelStatus_->setText(tr("SPI read failed."));
    } else if (spiReadProgress.wasCanceled()){
        labelStatus_->setText(tr("SPI read aborted by the user."));
    } else if (elapsedTime < 1000) {
        labelStatus_->setText(tr("SPI read completed in %1 ms.").arg(elapsedTime));
    } else {
        labelStatus_->setText(tr("SPI read completed in %1 s.").arg(locale_.toString(elapsedTime / 1000.0, 'f', 3)));
    }
    opCheck(tr("spi-read-op"), errcnt, errstr);  // The string "spi-read-op" should be translated to "SPI read"
}

// This function was expanded in version 3.0, in order to support transfers greater than 4096 bytes
void DeviceWindow::on_pushButtonWrite_clicked()
{
    quint8 channel = static_cast<quint8>(ui->comboBoxChannel->currentText().toUInt());
    size_t fragmentSizeLimit = evaluateSizeLimit();
    size_t bytesToWrite = write_.vector.size();
    size_t bytesProcessed = 0;
    QProgressDialog spiWriteProgress(tr("Performing SPI write..."), tr("Abort"), 0, static_cast<int>(bytesToWrite), this);  // Progress dialog implemented in version 3.0
    spiWriteProgress.setWindowTitle(tr("SPI Write"));
    spiWriteProgress.setWindowModality(Qt::WindowModal);
    spiWriteProgress.setMinimumDuration(500);  // The progress dialog should appear only if the operation takes more than 500 ms
    timer_->stop();  // Added!
    QElapsedTimer time;
    time.start();
    int errcnt = 0;
    QString errstr;
    cp2130_.selectCS(channel, errcnt, errstr);  // Enable the chip select corresponding to the selected channel, and disable any others
    usleep(100);  // Added!
    while (bytesProcessed < bytesToWrite) {
        if (spiWriteProgress.wasCanceled()) {  // If the user clicks "Abort"
            break;  // Abort the SPI write operation
        }
        size_t bytesRemaining = bytesToWrite - bytesProcessed;
        size_t fragmentSize = bytesRemaining > fragmentSizeLimit ? fragmentSizeLimit : bytesRemaining;
        cp2130_.spiWrite(write_.fragment(bytesProcessed, fragmentSize), epout_, errcnt, errstr);  // Write to the SPI bus
        if (errcnt > 0) {  // In case of error
            spiWriteProgress.cancel();  // Important!
            break;  // Abort the SPI write operation
        }
        bytesProcessed += fragmentSize;
        spiWriteProgress.setValue(static_cast<int>(bytesProcessed));
    }
    usleep(100);  // Wait 100 us, in order to prevent possible errors while disabling the chip select (workaround)
    cp2130_.disableCS(channel, errcnt, errstr);  // Disable the previously enabled chip select
    int elapsedTime = static_cast<int>(time.elapsed());  // Elapsed time in milliseconds
    timer_->start();  // Added!
    ui->lineEditRead->clear();
    if (errcnt > 0) {  // Update status bar
        labelStatus_->setText(tr("SPI write failed."));
    } else if (spiWriteProgress.wasCanceled()){
        labelStatus_->setText(tr("SPI write aborted by the user."));
    } else if (elapsedTime < 1000) {
        labelStatus_->setText(tr("SPI write completed in %1 ms.").arg(elapsedTime));
    } else {
        labelStatus_->setText(tr("SPI write completed in %1 s.").arg(locale_.toString(elapsedTime / 1000.0, 'f', 3)));
    }
    opCheck(tr("spi-write-op"), errcnt, errstr);  // The string "spi-write-op" should be translated to "SPI write"
}

// This function was expanded in version 3.0, similar to what was done with on_pushButtonRead_clicked() and on_pushButtonWrite_clicked(), but only to implement a progress dialog
void DeviceWindow::on_pushButtonWriteRead_clicked()
{
    quint8 channel = static_cast<quint8>(ui->comboBoxChannel->currentText().toUInt());
    size_t fragmentSizeLimit = evaluateSizeLimit();
    size_t bytesToWriteRead = write_.vector.size();
    size_t bytesProcessed = 0;
    QProgressDialog spiWriteReadProgress(tr("Performing SPI write and read..."), tr("Abort"), 0, static_cast<int>(bytesToWriteRead), this);  // Progress dialog implemented in version 3.0
    spiWriteReadProgress.setWindowTitle(tr("SPI Write/Read"));
    spiWriteReadProgress.setWindowModality(Qt::WindowModal);
    spiWriteReadProgress.setMinimumDuration(500);  // The progress dialog should appear only if the operation takes more than 500 ms
    Data read;
    timer_->stop();  // Added!
    QElapsedTimer time;
    time.start();
    int errcnt = 0;
    QString errstr;
    cp2130_.selectCS(channel, errcnt, errstr);  // Enable the chip select corresponding to the selected channel, and disable any others
    usleep(100);  // Added!
    while (bytesProcessed < bytesToWriteRead) {
        if (spiWriteReadProgress.wasCanceled()) {  // If the user clicks "Abort"
            break;  // Abort the SPI write and read operation
        }
        size_t bytesRemaining = bytesToWriteRead - bytesProcessed;
        size_t fragmentSize = bytesRemaining > fragmentSizeLimit ? fragmentSizeLimit : bytesRemaining;
        QVector<quint8> readFragment = cp2130_.spiWriteRead(write_.fragment(bytesProcessed, fragmentSize), epin_, epout_, errcnt, errstr);  // Write to and read from the SPI bus, simultaneously
        if (errcnt > 0) {  // In case of error
            spiWriteReadProgress.cancel();  // Important!
            break;  // Abort the SPI write and read operation
        }
        read.vector += readFragment;  // The returned fragment could be considered valid at this point
        bytesProcessed += fragmentSize;
        spiWriteReadProgress.setValue(static_cast<int>(bytesProcessed));
    }
    usleep(100);  // Wait 100 us, in order to prevent possible errors while disabling the chip select (workaround)
    cp2130_.disableCS(channel, errcnt, errstr);  // Disable the previously enabled chip select
    int elapsedTime = static_cast<int>(time.elapsed());  // Elapsed time in milliseconds
    timer_->start();  // Added!
    ui->lineEditRead->setText(read.toHexadecimal());  // At least, a partial result should be shown if an error occurs
    if (errcnt > 0) {  // Update status bar
        labelStatus_->setText(tr("SPI write and read failed."));
    } else if (spiWriteReadProgress.wasCanceled()){
        labelStatus_->setText(tr("SPI write and read aborted by the user."));
    } else if (elapsedTime < 1000) {
        labelStatus_->setText(tr("SPI write and read completed in %1 ms.").arg(elapsedTime));
    } else {
        labelStatus_->setText(tr("SPI write and read completed in %1 s.").arg(locale_.toString(elapsedTime / 1000.0, 'f', 3)));
    }
    opCheck(tr("spi-write-read-op"), errcnt, errstr);  // The string "spi-write-read-op" should be translated to "SPI write and read"
}
```

**Comment by samuelfmlourenco** [comment]
The previous fix completely eliminated the previous failure mode, so it no longer fails during the first bulk transfer. However, SPI write transfers still fail when the frequency is set to 750 kHz and below. However, I was able to do 100 transfers of 256 KiB each, at 93.8 kHz, without issues.

In a nutshell, although it is impossible to prevent failures completely, the issue is greatly mitigated by the proposed changes above. Failures in the middle of SPI transfers still happen without any visible cause. The last proposed change will be put into effect on the next release.

**Comment by samuelfmlourenco** [comment]
Although this issue was not fixed, the proposed corrections will minimize its occurrence. The code, as applied, follows:

`const float TIME_LIMIT = 100;  // Soft time limit per partial transfer, in milliseconds`

```
// This function was expanded in version 3.0, in order to support transfers greater than 4096 bytes
void DeviceWindow::on_pushButtonRead_clicked()
{
    quint8 channel = static_cast<quint8>(ui->comboBoxChannel->currentText().toUInt());
    size_t fragmentSizeLimit = calculateSizeLimit();
    size_t bytesToRead = static_cast<size_t>(ui->spinBoxBytesToRead->value());
    size_t bytesProcessed = 0;
    QProgressDialog spiReadProgress(tr("Performing SPI read..."), tr("Abort"), 0, static_cast<int>(bytesToRead), this);  // Progress dialog implemented in version 3.0
    spiReadProgress.setWindowTitle(tr("SPI Read"));
    spiReadProgress.setWindowModality(Qt::WindowModal);
    spiReadProgress.setMinimumDuration(500);  // The progress dialog should appear only if the operation takes more than 500 ms
    Data read;
    timer_->stop();  // The update timer is now stopped during SPI transfers (fix implemented in version 3.1)
    QElapsedTimer time;
    time.start();
    int errcnt = 0;
    QString errstr;
    cp2130_.selectCS(channel, errcnt, errstr);  // Enable the chip select corresponding to the selected channel, and disable any others
    usleep(100);  // Wait 100 us, in order to prevent possible errors after enabling the chip select (workaround implemented in version 3.1)
    while (bytesProcessed < bytesToRead) {
        if (spiReadProgress.wasCanceled()) {  // If the user clicks "Abort"
            break;  // Abort the SPI read operation
        }
        size_t bytesRemaining = bytesToRead - bytesProcessed;
        size_t fragmentSize = bytesRemaining > fragmentSizeLimit ? fragmentSizeLimit : bytesRemaining;
        QVector<quint8> readFragment = cp2130_.spiRead(static_cast<quint32>(fragmentSize), epin_, epout_, errcnt, errstr);  // Read from the SPI bus
        if (errcnt > 0) {  // In case of error
            spiReadProgress.cancel();  // Important!
            break;  // Abort the SPI read operation
        }
        read.vector += readFragment;  // The returned fragment could be considered valid at this point
        bytesProcessed += fragmentSize;
        spiReadProgress.setValue(static_cast<int>(bytesProcessed));
    }
    usleep(100);  // Wait 100 us, in order to prevent possible errors while disabling the chip select (workaround)
    cp2130_.disableCS(channel, errcnt, errstr);  // Disable the previously enabled chip select
    qint64 elapsedTime = time.elapsed();  // Elapsed time in milliseconds
    timer_->start();  // Restart the timer
    ui->lineEditRead->setText(read.toHexadecimal());  // At least, a partial result should be shown in case of error
    if (errcnt > 0) {  // Update status bar
        labelStatus_->setText(tr("SPI read failed."));
    } else if (spiReadProgress.wasCanceled()){
        labelStatus_->setText(tr("SPI read aborted by the user."));
    } else if (elapsedTime < 1000) {
        labelStatus_->setText(tr("SPI read completed in %1 ms.").arg(elapsedTime));
    } else {
        labelStatus_->setText(tr("SPI read completed in %1 s.").arg(locale_.toString(elapsedTime / 1000.0, 'f', 3)));
    }
    opCheck(tr("spi-read-op"), errcnt, errstr);  // The string "spi-read-op" should be translated to "SPI read"
}

// This function was expanded in version 3.0, in order to support transfers greater than 4096 bytes
void DeviceWindow::on_pushButtonWrite_clicked()
{
    quint8 channel = static_cast<quint8>(ui->comboBoxChannel->currentText().toUInt());
    size_t fragmentSizeLimit = calculateSizeLimit();
    size_t bytesToWrite = write_.vector.size();
    size_t bytesProcessed = 0;
    QProgressDialog spiWriteProgress(tr("Performing SPI write..."), tr("Abort"), 0, static_cast<int>(bytesToWrite), this);  // Progress dialog implemented in version 3.0
    spiWriteProgress.setWindowTitle(tr("SPI Write"));
    spiWriteProgress.setWindowModality(Qt::WindowModal);
    spiWriteProgress.setMinimumDuration(500);  // The progress dialog should appear only if the operation takes more than 500 ms
    timer_->stop();  // The update timer is now stopped during SPI transfers (fix implemented in version 3.1)
    QElapsedTimer time;
    time.start();
    int errcnt = 0;
    QString errstr;
    cp2130_.selectCS(channel, errcnt, errstr);  // Enable the chip select corresponding to the selected channel, and disable any others
    usleep(100);  // Wait 100 us, in order to prevent possible errors after enabling the chip select (workaround implemented in version 3.1)
    while (bytesProcessed < bytesToWrite) {
        if (spiWriteProgress.wasCanceled()) {  // If the user clicks "Abort"
            break;  // Abort the SPI write operation
        }
        size_t bytesRemaining = bytesToWrite - bytesProcessed;
        size_t fragmentSize = bytesRemaining > fragmentSizeLimit ? fragmentSizeLimit : bytesRemaining;
        cp2130_.spiWrite(write_.fragment(bytesProcessed, fragmentSize), epout_, errcnt, errstr);  // Write to the SPI bus
        if (errcnt > 0) {  // In case of error
            spiWriteProgress.cancel();  // Important!
            break;  // Abort the SPI write operation
        }
        bytesProcessed += fragmentSize;
        spiWriteProgress.setValue(static_cast<int>(bytesProcessed));
    }
    usleep(100);  // Wait 100 us, in order to prevent possible errors while disabling the chip select (workaround)
    cp2130_.disableCS(channel, errcnt, errstr);  // Disable the previously enabled chip select
    qint64 elapsedTime = time.elapsed();  // Elapsed time in milliseconds
    timer_->start();  // Restart the timer
    ui->lineEditRead->clear();
    if (errcnt > 0) {  // Update status bar
        labelStatus_->setText(tr("SPI write failed."));
    } else if (spiWriteProgress.wasCanceled()){
        labelStatus_->setText(tr("SPI write aborted by the user."));
    } else if (elapsedTime < 1000) {
        labelStatus_->setText(tr("SPI write completed in %1 ms.").arg(elapsedTime));
    } else {
        labelStatus_->setText(tr("SPI write completed in %1 s.").arg(locale_.toString(elapsedTime / 1000.0, 'f', 3)));
    }
    opCheck(tr("spi-write-op"), errcnt, errstr);  // The string "spi-write-op" should be translated to "SPI write"
}

// This function was expanded in version 3.0, similar to what was done with on_pushButtonRead_clicked() and on_pushButtonWrite_clicked(), but only to implement a progress dialog
void DeviceWindow::on_pushButtonWriteRead_clicked()
{
    quint8 channel = static_cast<quint8>(ui->comboBoxChannel->currentText().toUInt());
    size_t fragmentSizeLimit = calculateSizeLimit();
    size_t bytesToWriteRead = write_.vector.size();
    size_t bytesProcessed = 0;
    QProgressDialog spiWriteReadProgress(tr("Performing SPI write and read..."), tr("Abort"), 0, static_cast<int>(bytesToWriteRead), this);  // Progress dialog implemented in version 3.0
    spiWriteReadProgress.setWindowTitle(tr("SPI Write/Read"));
    spiWriteReadProgress.setWindowModality(Qt::WindowModal);
    spiWriteReadProgress.setMinimumDuration(500);  // The progress dialog should appear only if the operation takes more than 500 ms
    Data read;
    timer_->stop();  // The update timer is now stopped during SPI transfers (fix implemented in version 3.1)
    QElapsedTimer time;
    time.start();
    int errcnt = 0;
    QString errstr;
    cp2130_.selectCS(channel, errcnt, errstr);  // Enable the chip select corresponding to the selected channel, and disable any others
    usleep(100);  // Wait 100 us, in order to prevent possible errors after enabling the chip select (workaround implemented in version 3.1)
    while (bytesProcessed < bytesToWriteRead) {
        if (spiWriteReadProgress.wasCanceled()) {  // If the user clicks "Abort"
            break;  // Abort the SPI write and read operation
        }
        size_t bytesRemaining = bytesToWriteRead - bytesProcessed;
        size_t fragmentSize = bytesRemaining > fragmentSizeLimit ? fragmentSizeLimit : bytesRemaining;
        QVector<quint8> readFragment = cp2130_.spiWriteRead(write_.fragment(bytesProcessed, fragmentSize), epin_, epout_, errcnt, errstr);  // Write to and read from the SPI bus, simultaneously
        if (errcnt > 0) {  // In case of error
            spiWriteReadProgress.cancel();  // Important!
            break;  // Abort the SPI write and read operation
        }
        read.vector += readFragment;  // The returned fragment could be considered valid at this point
        bytesProcessed += fragmentSize;
        spiWriteReadProgress.setValue(static_cast<int>(bytesProcessed));
    }
    usleep(100);  // Wait 100 us, in order to prevent possible errors while disabling the chip select (workaround)
    cp2130_.disableCS(channel, errcnt, errstr);  // Disable the previously enabled chip select
    qint64 elapsedTime = time.elapsed();  // Elapsed time in milliseconds
    timer_->start();  // Restart the timer
    ui->lineEditRead->setText(read.toHexadecimal());  // At least, a partial result should be shown if an error occurs
    if (errcnt > 0) {  // Update status bar
        labelStatus_->setText(tr("SPI write and read failed."));
    } else if (spiWriteReadProgress.wasCanceled()){
        labelStatus_->setText(tr("SPI write and read aborted by the user."));
    } else if (elapsedTime < 1000) {
        labelStatus_->setText(tr("SPI write and read completed in %1 ms.").arg(elapsedTime));
    } else {
        labelStatus_->setText(tr("SPI write and read completed in %1 s.").arg(locale_.toString(elapsedTime / 1000.0, 'f', 3)));
    }
    opCheck(tr("spi-write-read-op"), errcnt, errstr);  // The string "spi-write-read-op" should be translated to "SPI write and read"
}
```

```
// Calculates the optimal fragment size limit based on the parameters of the currently selected channel (implemented in version 3.1, to replace evaluateSizeLimit())
size_t DeviceWindow::calculateSizeLimit()
{
    QString channelName = ui->comboBoxChannel->currentText();
    float preDeassertDelay = spiDelaysMap_[channelName].prdasten == true ? spiDelaysMap_[channelName].prdastdly / 100.0 : 0;  // Pre-deassert delay in milliseconds
    float postAssertDelay = spiDelaysMap_[channelName].pstasten == true ? spiDelaysMap_[channelName].pstastdly / 100.0 : 0;  // Post-assert delay in milliseconds
    float interByteDelay = spiDelaysMap_[channelName].itbyten == true ? spiDelaysMap_[channelName].itbytdly / 100.0 : 0;  // Inter-byte delay in milliseconds
    float timePerByte = std::pow(2, ui->comboBoxFrequency->currentIndex()) / 1500;  // Time duration for each byte, in milliseconds
    size_t sizeLimit = static_cast<size_t>((TIME_LIMIT + interByteDelay - preDeassertDelay - postAssertDelay) / (timePerByte + interByteDelay));
    return sizeLimit == 0 ? 1 : sizeLimit;  // The size limit cannot be zero
}
```

Note that calculateSizeLimit() replaces evaluateSizeLimit(). Although the returned size does not follow discrete and easily predictable values, this new function is far more optimized and also accounts for increased SPI delays.
